pkg/container/ring.go

```diff
@@ -89,12 +110,30 @@ func (r *Ring) Cap() uint64 {
 }
 
 // Write writes the given event into the ring buffer in the next available
-// writing block.
+// writing block. The entry must not be nil, otherwise readFrom will block when
+// reading back this event.
 func (r *Ring) Write(entry *v1.Event) {
+	// We need to lock the notification mutex when updating r.write, otherwise
+	// there is a race condition where a readFrom goroutine goes to sleep
+	// after we sent out the notification.
+	// This lock is only shared with other readFrom instances that are about
+	// to go to sleep, so contention should be low. Notably, readers which are
+	// far away from the current write pointer will still be able to make
+	// progress concurrently.
+
+	r.notifyMu.Lock()
+
 	write := atomic.AddUint64(&r.write, 1)
 	writeIdx := (write - 1) & r.mask
-	r.data[writeIdx] = entry
-	r.cond.Broadcast()
+	r.dataStoreAtomic(writeIdx, entry)
+
+	// notify any sleeping readers
+	if r.notifyCh != nil {
+		close(r.notifyCh)
```

Comment:
Is it allowed to have multiple writers? Because if yes, then we have a race here and the channel could be closed twice thus leading to panic.

Reply:
Yeah... also we don't allow multiple writers so far :)